### ISSUE TITLE
prm, moe_lora_routing, unlearn and ppo drop the optimizer and schedule: the default config's cosine/warmup/weight-decay never reach them, and batch_size: auto becomes 1

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#805]
## The defect

Four trainers build their optimizer settings without most of `TrainingConfig`. Because the schema defaults are not the HF/TRL defaults (`optimizer=adamw_torch`, `scheduler=cosine`, `warmup_ratio=0.03`, `weight_decay=0.01`, `save_steps=100`, `batch_size=auto`), this changes even a config that sets none of them.

### prm and moe_lora_routing

`trainer/prm.py:323-338` and `trainer/mole_routing.py:361-374` construct `TrainingArguments` with epochs, batch, accumulation, lr, logging/save steps, and nothing else. Omitted: `optim`, `lr_scheduler_type`, `warmup_steps`, `weight_decay`, `max_grad_norm`, `gradient_checkpointing`; MoLE also omits `bf16`/`fp16` (PRM passes them). With transformers 5.17 the omitted values fall back to:

```

optim = adamw_torch_fused   lr_scheduler_type = linear   weight_decay = 0.0

max_grad_norm = 1.0         warmup_steps = 0             gradient_checkpointing = False

```

`batch_size: auto`, the schema default, is not estimated as in the other trainers; it silently becomes 1 (`prm.py:318-321`, `mole_routing.py:357-360`):

```python

if isinstance(tcfg.batch_size, bool) or not isinstance(tcfg.batch_size, int):

    bs = 1

```

### unlearn

`trainer/unlearn.py` has its own loop. `setup()` (`:222-224`) builds `torch.optim.AdamW(params, lr=lr)` with torch's `weight_decay=0.01` whatever the config says, and no scheduler, warmup, clipping or optimizer choice. `train()` (`:277-305`) takes one forget example per optimizer step, so `batch_size` and `gradient_accumulation_steps` (default 4) are ignored. Sequences are cut at a module constant `_MAX_LENGTH = 256` (`:39`), not `data.max_length` (default 2048). The base loads through `live_eval.load_model_and_tokenizer` without `quantization` (default `4bit`), and the LoRA config (`:184-190`) passes only `r`, `alpha`, `dropout`, so `target_modules`, `use_dora`, `use_rslora` and `rank_pattern` are ignored (PEFT picks its per-architecture default targets).

### ppo

`trainer/ppo.py:157-163` starts `ppo_kwargs` with output dir, batch, accumulation, lr and seed; the capability probes after it add PPO-specific fields only. trl 0.29.1's `trl.experimental.ppo.PPOConfig` then supplies:

```

optim = adamw_torch_fused   lr_scheduler_type = linear   weight_decay = 0.0

max_grad_norm = 1.0         warmup_steps = 0             save_steps = 500   logging_steps = 10

```

so the default config's `save_steps: 100` becomes 500 and `cosine` becomes `linear`.

**Coordination:** #721 (open, PRs #737 and #741) covers PPO's `kl_coef` and `num_train_epochs`. Neither PR touches the fields above; I checked both diffs. They are additional and could land in whichever PR wins there, or after it.

## Reproduction

All accepted, current `main`:

```

ACCEPT prm batch_size= auto optimizer= adafactor scheduler= cosine wd= 0.1 mgn= 0.3 dora= True

ACCEPT moe_lora_routing batch_size= auto optimizer= adafactor scheduler= cosine wd= 0.1 mgn= 0.3 dora= True

ACCEPT unlearn batch_size= auto optimizer= adafactor scheduler= cosine wd= 0.1 mgn= 0.3 dora= True

```

None of those values reaches training on these three tasks.

## Why it matters

A PRM trained with the default config gets a linear schedule with no warmup and no weight decay at batch 1, where the same YAML on SFT gets cosine, 3% warmup and decay 0.01 at an estimated batch. Nothing is printed. Unlearning is where step size matters most, since NPO/RMU push the model away from data; there the user cannot clip, warm up or schedule, and their `lora.target_modules` is replaced.

## Proposed fix

- prm / moe_lora_routing: pass the same fields every other HF-Trainer wrapper passes (see `trainer/dpo.py`'s `DPOConfig(...)`), and resolve `batch_size: auto` through `estimate_batch_size` or refuse `auto` at load for these tasks. A shared `training_arguments_kwargs(tcfg, ...)` helper, like `training_seed_kwargs` for #353, keeps the next trainer from repeating this.

- ppo: add the same fields to `ppo_kwargs`, probed by capability like the existing ones.

- unlearn: honour `weight_decay`, `max_grad_norm` (`clip_grad_norm_`), a scheduler with warmup, `gradient_accumulation_steps`, `data.max_length`, `quantization` and the full LoRA config, or reject each unsupported one for `task: unlearn` at load.

## Acceptance criteria

- [ ] For prm and moe_lora_routing, a test builds the trainer on a tiny CPU model with `scheduler: cosine, warmup_ratio: 0.1, weight_decay: 0.1, max_grad_norm: 0.3, optimizer: adafactor` and asserts each value on `trainer.args`. Fails today.

- [ ] `batch_size: auto` on prm / moe_lora_routing is either estimated or refused, not silently 1.

- [ ] PPO: the same assertions against `PPOConfig`, written so a fall-back to trl defaults cannot pass (non-default values).

- [ ] unlearn: a test that a configured `weight_decay` reaches the optimizer's param groups and `max_grad_norm` clips; unsupported fields refused at load.

Related: #745 (`loraplus_lr_ratio`), same shape.

---

**Related, filed the same day:**

- #795 — The default quantization: 4bit is ignored by six trainers (distill, classifier/reranker/cross_encoder, prm,...
- #802 — loss_spike_recovery and grad_accum_auto_tune reach the training callback on SFT only: 15 trainers build it ...

**Comment by MakazhanAlpamys** [comment]
**Claimed by @Oxygen56**, working in #868.

Recorded here rather than only in the pull request, because the assignee field silently refused: GitHub accepts an assignee only if they have push access or have commented on **this issue**, and `gh issue edit --add-assignee` exits 0 either way. That is how 11 claims ended up invisible today, including this one.

So the `claimed` label plus this comment is the registry. If you are looking for unclaimed work, trust the label and the comments, not an empty assignee field.

**Comment by MakazhanAlpamys** [comment]
Re-checked on `main` `99d553f5` while verifying a code-review finding. The PPO part of this issue still reproduces with the real `trl.experimental.ppo.PPOTrainer` built on CPU. With `optimizer: adafactor`, `scheduler: cosine`, `warmup_ratio: 0.1`, `weight_decay: 0.05`, `max_grad_norm: 0.3`, `save_steps: 7`, `logging_steps: 3`, the trainer gets `optim=adamw_torch_fused`, `linear`, `warmup_steps=0`, `weight_decay=0.0`, `max_grad_norm=1.0`, `save_steps=500`, `logging_steps=10`, and the optimizer it builds is `torch.optim.AdamW` with weight decay 0.0.

**Three more PPO settings are dropped that this issue does not list.** #868's PPO hunk does not add them either; it adds `warmup_steps`, `weight_decay`, `max_grad_norm`, `optim`, `lr_scheduler_type`, `logging_steps`, `save_steps`, `bf16`/`fp16` and `gradient_checkpointing`.

1. **`--deepspeed` never reaches PPO.** `soup train --deepspeed` hands the wrapper its config path, but `ppo_kwargs` (`src/soup_cli/trainer/ppo.py:205-211`) has no `deepspeed` key, so the run is not under DeepSpeed at all. `ppo.py:373-376` still attaches the empty-param-group DeepSpeed guard because `self.deepspeed_config` is set. The other 16 wrappers pass `deepspeed=self.deepspeed_config`.
2. **The response length cannot be configured.** `response_length` stays at trl's default of 53 tokens whatever `data.max_length` says. GRPO maps `data.max_length` to `max_completion_length` (`src/soup_cli/trainer/grpo.py:480`); PPO forwards nothing to `response_length`.
3. **`save_total_limit` is `None`**, so PPO keeps every checkpoint. Every other wrapper built on a Hugging Face trainer passes `save_total_limit=3` (TTS inherits it from SFT).

```python
import json
from pathlib import Path

from soup_cli.config.loader import load_config_from_string
from soup_cli.data.loader import load_dataset
from soup_cli.trainer.ppo import PPOTrainerWrapper

tiny = "hf-internal-testing/tiny-random-LlamaForCausalLM"
Path("prompts.jsonl").write_text("\n".join(json.dumps(
    {"messages": [{"role": "user", "content": f"Say something about {i}."}]}) for i in range(8)))
Path("ds.json").write_text('{"zero_optimization": {"stage": 3}, "train_batch_size": "auto"}')
cfg = load_config_from_string(
    f"base: {tiny}\ntask: ppo\n"
    "data: {train: prompts.jsonl, format: chatml, val_split: 0.0, max_length: 512}\n"
    f"training: {{reward_model: {tiny}, quantization: none, batch_size: 2}}\n"
    "output: ./out\n")
wrapper = PPOTrainerWrapper(cfg, device="cpu", deepspeed_config="ds.json")  # as `--deepspeed` passes it
wrapper.setup(load_dataset(cfg.data))
args = wrapper.trainer.args
print("args.deepspeed:", args.deepspeed)
print("accelerator deepspeed_plugin:", wrapper.trainer.accelerator.state.deepspeed_plugin)
print("empty-group DeepSpeed guard attached:", getattr(wrapper.trainer, "_soup_empty_group_guard", False))
print("data.max_length:", cfg.data.max_length, "| args.response_length:", args.response_length)
print("args.save_total_limit:", args.save_total_limit)
```

```
args.deepspeed: None
accelerator deepspeed_plugin: None
empty-group DeepSpeed guard attached: True
data.max_length: 512 | args.response_length: 53
args.save_total_limit: None
```

Suggested additions to the acceptance list:

- [ ] `task: ppo` with `--deepspeed <config>`: `trainer.args.deepspeed` is that config, or `--deepspeed` is refused for `task: ppo` with a message naming the task. Either way, the guard is not attached to a run that has no DeepSpeed engine.
- [ ] `trainer.args.response_length` follows a config value (`data.max_length`, or a dedicated field), asserted with a value other than 53.
- [ ] `trainer.args.save_total_limit == 3`, as on the other Hugging Face-based wrappers.
